float32 angle
float64 timestamp